File: packages/vscode-extension/src/whats-new.js
/**
 * "What's New" — surface the cc CLI's own release notes (`cc changelog
 * --json`, shipped offline with the npm package since 0.162.151) inside the
 * IDE, and nudge once after an upgrade is detected. Pure logic (no `vscode`):
 * extension.js wires the command + toast.
 */

const fs = require("node:fs");
const path = require("node:path");

const MIN_CHANGELOG_CLI = "0.162.151"; // first cc with `cc changelog`

/** argv for the CLI call behind the panel. */
function buildChangelogArgs({ limit = 5 } = {}) {
  return ["changelog", "-n", String(limit), "--json"];
}

/** Parse `cc changelog --json` stdout → releases[], or null when unusable
 * (old CLI printing an "unknown command" error, garbage, etc.). */
function parseChangelogJson(text) {
  try {
    const data = JSON.parse(String(text || ""));
    return Array.isArray(data?.releases) ? data.releases : null;
  } catch {
    return null;
  }
}

/**
 * Prefer usable CLI output, but let an empty (rather than malformed) JSON
 * result fall back to the artifact bundled in the same installation.
 */
function chooseChangelogReleases(parsed, bundled) {
  if (Array.isArray(parsed) && parsed.length > 0) return parsed;
  if (Array.isArray(bundled) && bundled.length > 0) return bundled;
  return Array.isArray(parsed)
    ? parsed
    : Array.isArray(bundled)
      ? bundled
      : null;
}

function releaseVersions(release) {
  return [
    release?.cliVersion,
    ...(Array.isArray(release?.cliVersions) ? release.cliVersions : []),
  ].filter(Boolean);
}

/**
 * Detect a non-empty but stale changelog. Older published CLI packages can
 * report a newer `cc --version` than any release in their bundled artifact.
 * Never present that data as if it described the installed build.
 */
function reconcileChangelogReleases(releases, { installed } = {}) {
  const list = Array.isArray(releases) ? releases : [];
  const versions = [...new Set(list.flatMap(releaseVersions))];
  if (!installed || versions.includes(installed) || versions.length === 0) {
    return { releases: list, stale: false };
  }
  const { compareVersions } = require("./version-check");
  const latest = versions.reduce((best, version) =>
    compareVersions(version, best) > 0 ? version : best,
  );
  if (compareVersions(installed, latest) <= 0) {
    return { releases: list, stale: false };
  }

  return {
    stale: true,
    releases: [
      {
        productVersion: null,
        cliVersion: installed,
        cliVersions: [installed],
        date: null,
        title: "Installed CLI build — matching release notes were not bundled",
        sections: [
          {
            heading: "Release-note data is out of date",
            body:
              `- Installed CLI: \`${installed}\`\n` +
              `- Newest bundled notes: \`${latest}\`\n\n` +
              "Older bundled notes follow below. Reinstall or upgrade the CLI " +
              "after a corrected package is published.",
          },
        ],
      },
      ...list,
    ],
  };
}

/**
 * Read the changelog artifact shipped inside the installed CLI package.
 *
 * Older Windows installations can have a parent-level `CHANGELOG.md` (the
 * Node.js distribution ships one) that makes `cc changelog --json` return an
 * empty release list. The npm package still contains the authoritative
 * `src/data/changelog.json`, so the IDE can recover without waiting for a CLI
 * upgrade or requiring network access.
 */
function loadBundledChangelog({
  command = "cc",
  env = process.env,
  fsApi = fs,
  pathApi = path,
} = {}) {
  const commandText = typeof command === "string" ? command.trim() : "";
  const pathEntries = String(env?.PATH || "")
    .split(pathApi.delimiter)
    .filter(Boolean);
  const commandCandidates = pathApi.isAbsolute(commandText)
    ? [commandText]
    : [
        commandText,
        ...pathEntries.flatMap((entry) =>
          process.platform === "win32"
            ? [
                pathApi.join(entry, `${commandText}.cmd`),
                pathApi.join(entry, commandText),
              ]
            : [pathApi.join(entry, commandText)],
        ),
      ];
  const packageRoots = [];
  for (const executable of commandCandidates) {
    if (!pathApi.isAbsolute(executable)) continue;
    const dir = pathApi.dirname(executable);
    packageRoots.push(
      pathApi.join(dir, "node_modules", "chainlesschain"),
      pathApi.join(dir, "..", "node_modules", "chainlesschain"),
    );
  }
  for (const root of [...new Set(packageRoots)]) {
    const dataPath = pathApi.join(root, "src", "data", "changelog.json");
    try {
      if (!fsApi.existsSync(dataPath)) continue;
      const data = JSON.parse(fsApi.readFileSync(dataPath, "utf8"));
      if (Array.isArray(data?.releases) && data.releases.length > 0)
        return data.releases;
    } catch {
      // A broken optional fallback must not hide the normal CLI error path.
    }
  }
  return null;
}

/**
 * Render releases into the markdown document the IDE shows. The section
 * bodies come straight from the package's own curated changelog, so they are
 * passed through as-is (they already ARE markdown).
 */
function changelogToMarkdown(releases, { installed } = {}) {
  const lines = ["# cc CLI — What's New", ""];
  for (const rel of releases || []) {
    const meta = [
      rel.productVersion ? `product ${rel.productVersion}` : null,
      rel.date || null,
    ]
      .filter(Boolean)
      .join(", ");
    const mark =
      installed && rel.cliVersion === installed ? " ← installed" : "";
    lines.push(`## ${rel.cliVersion}${meta ? ` (${meta})` : ""}${mark}`, "");
    if (rel.title) lines.push(rel.title, "");
    for (const s of rel.sections || []) {
      if (s.heading) lines.push(`### ${s.heading}`, "");
      if (s.body) lines.push(s.body.replace(/^>\s?/gm, ""), "");
    }
  }
  return lines.join("\n").replace(/\n{3,}/g, "\n\n");
}

/**
 * One-shot upgrade nudge decision. `prev` is the version remembered from the
 * last activation (null on first run — store silently, no toast: nagging a
 * fresh install with "you upgraded!" is noise). Returns null or the toast
 * payload.
 */
function upgradeNudge(prev, current) {
  if (!current) return null; // cc missing/unparsable — nothing to remember
  if (!prev || prev === current) return null;
  // Only nudge on an actual UPGRADE. A downgrade (npm i -g chainlesschain@older)
  // shouldn't say "updated" and offer notes the installed CLI may not even have
  // (cc changelog needs >= 0.162.151, so the button would then just error).
  const { compareVersions } = require("./version-check");
  if (compareVersions(current, prev) <= 0) return null;
  return {
    message: `cc CLI updated: ${prev} → ${current}`,
    button: "What's New",
  };
}

module.exports = {
  MIN_CHANGELOG_CLI,
  buildChangelogArgs,
  parseChangelogJson,
  chooseChangelogReleases,
  reconcileChangelogReleases,
  loadBundledChangelog,
  changelogToMarkdown,
  upgradeNudge,
};

File: packages/jetbrains-plugin/src/main/java/com/chainlesschain/ide/WhatsNew.java
package com.chainlesschain.ide;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

/**
 * "What's New" — surface the cc CLI's own release notes ({@code cc changelog
 * --json}, shipped offline with the npm package since 0.162.151) inside the
 * IDE. The Java twin of the VS Code extension's whats-new.js: pure arg
 * builder + tolerant parser + text renderer; the glue (ShowWhatsNewAction)
 * drives {@code AgentChatSession.runCapture} + a dialog around them.
 */
public final class WhatsNew {

    private WhatsNew() {}

    /** First cc release that ships {@code cc changelog}. */
    public static final String MIN_CHANGELOG_CLI = "0.162.151";

    /** One release entry from {@code cc changelog --json}. */
    public static final class Release {
        public final String cliVersion;
        public final String productVersion; // nullable
        public final String date;           // nullable
        public final String title;          // nullable
        /** [heading (nullable), body (nullable)] pairs in document order. */
        public final List<String[]> sections;

        Release(String cliVersion, String productVersion, String date, String title,
                List<String[]> sections) {
            this.cliVersion = cliVersion;
            this.productVersion = productVersion;
            this.date = date;
            this.title = title;
            this.sections = sections;
        }
    }

    /** {@code cc changelog -n <limit> --json} — the panel's argv. */
    public static List<String> buildChangelogArgs(int limit) {
        return new ArrayList<String>(Arrays.asList(
                "changelog", "-n", String.valueOf(limit <= 0 ? 5 : limit), "--json"));
    }

    /** Does the installed cc ship {@code cc changelog}? (raw version output ok) */
    public static boolean supportsChangelog(String installedRaw) {
        String installed = CliVersionCheck.parseVersion(installedRaw);
        return installed != null
                && CliVersionCheck.compare(installed, MIN_CHANGELOG_CLI) >= 0;
    }

    /**
     * Tolerant parse of {@code cc changelog --json} stdout → releases (empty on
     * any mismatch — an old CLI printing an "unknown command" error, garbage…).
     */
    public static List<Release> parseChangelogJson(String stdout) {
        List<Release> out = new ArrayList<Release>();
        Map<String, Object> data;
        try {
            data = MiniJson.parseObject(stdout == null ? "" : stdout.trim());
        } catch (RuntimeException e) {
            return out;
        }
        Object releases = data.get("releases");
        if (!(releases instanceof List)) return out;
        for (Object row : (List<?>) releases) {
            if (!(row instanceof Map)) continue;
            Map<?, ?> r = (Map<?, ?>) row;
            Object cli = r.get("cliVersion");
            if (!(cli instanceof String) || ((String) cli).isEmpty()) continue;
            List<String[]> sections = new ArrayList<String[]>();
            Object secs = r.get("sections");
            if (secs instanceof List) {
                for (Object so : (List<?>) secs) {
                    if (!(so instanceof Map)) continue;
                    Map<?, ?> s = (Map<?, ?>) so;
                    sections.add(new String[] {
                            asStringOrNull(s.get("heading")),
                            asStringOrNull(s.get("body")) });
                }
            }
            out.add(new Release(
                    (String) cli,
                    asStringOrNull(r.get("productVersion")),
                    asStringOrNull(r.get("date")),
                    asStringOrNull(r.get("title")),
                    sections));
        }
        return out;
    }

    /**
     * Render releases into the text document the dialog shows — the same
     * markdown VS Code's changelogToMarkdown produces (the section bodies come
     * straight from the package's own curated changelog, so they already ARE
     * markdown and read fine as plain text). {@code installed} (nullable) marks
     * the user's version with "← installed".
     */
    public static String changelogToText(List<Release> releases, String installed) {
        StringBuilder sb = new StringBuilder("# cc CLI — What's New\n\n");
        for (Release rel : releases == null ? new ArrayList<Release>() : releases) {
            StringBuilder meta = new StringBuilder();
            if (rel.productVersion != null && !rel.productVersion.isEmpty()) {
                meta.append("product ").append(rel.productVersion);
            }
            if (rel.date != null && !rel.date.isEmpty()) {
                if (meta.length() > 0) meta.append(", ");
                meta.append(rel.date);
            }
            String mark = installed != null && installed.equals(rel.cliVersion)
                    ? " ← installed" : "";
            sb.append("## ").append(rel.cliVersion)
              .append(meta.length() > 0 ? " (" + meta + ")" : "")
              .append(mark).append("\n\n");
            if (rel.title != null && !rel.title.isEmpty()) {
                sb.append(rel.title).append("\n\n");
            }
            for (String[] s : rel.sections) {
                if (s[0] != null && !s[0].isEmpty()) sb.append("### ").append(s[0]).append("\n\n");
                if (s[1] != null && !s[1].isEmpty()) {
                    // Strip the changelog's blockquote markers — plain text dialog.
                    sb.append(s[1].replaceAll("(?m)^>\\s?", "")).append("\n\n");
                }
            }
        }
        return sb.toString().replaceAll("\n{3,}", "\n\n");
    }

    private static String asStringOrNull(Object v) {
        return v instanceof String && !((String) v).isEmpty() ? (String) v : null;
    }
}
